hubblestack/daemon.py

```diff
@@ -466,6 +469,45 @@ def parse_args():
                         help='Any arguments necessary for a single function run')
     return vars(parser.parse_args())
 
+def check_pidfile(kill_other=False):
+    '''
+    Check to see if there's already a pidfile. If so, check to see if the
+    indicated process is alive and is Hubble.
+
+    kill_other
+        Default false, if set to true, attempt to kill detected running Hubble
+        processes; otherwise exit with an error.
+
+    '''
+    pidfile = __opts__['pidfile']
+    if os.path.isfile(pidfile):
+        with open(pidfile, 'r') as f:
+            xpid = f.readline().strip()
+            try:
+                xpid = int(xpid)
+            except:
+                xpid = 0
+                log.warn('unable to parse pid="{pid}" in pidfile={file}'.format(pid=xpid,file=pidfile))
+            if xpid:
+                log.warn('pidfile={file} exists and contains pid={pid}'.format(file=pidfile, pid=xpid))
+                if os.path.isdir("/proc/{pid}".format(pid=xpid)):
+                    with open("/proc/{pid}/cmdline".format(pid=xpid),'r') as f2:
+                        cmdline = f2.readline().strip().strip('\x00').replace('\x00',' ')
+                        if 'hubble' in cmdline:
+                            if kill_other:
+                                log.warn("process seems to still be alive and is hubble, attempting to shutdown")
+                                os.kill(int(xpid), signal.SIGTERM)
+                                time.sleep(1)
+                                if os.path.isdir("/proc/{pid}".format(pid=xpid)):
+                                    log.error("failed to shutdown process successfully; abnormal program exit")
+                                    exit(1)
+                                else:
+                                    log.info("shutdown seems to have succeeded, proceeding with startup")
```

Comment:
FYI, error will be sent to a returner with a sourcetype of _hubble_log_. I think it would be useful to detect and see in that data indications that this action (killing a concurrent hubble process) is happening.

It also would probably be useful to include some details with the action, like the _pid_.

Reply:
I'm not convinced the killed process would invoke any sort of return. A TERM shutdown seems to be caught as a single to do a normal shutdown (as if it had come from sysv or systemd). And in the case where it can't send a TERM or it doesn't work and the new process shuts down in stead, it will not have gotten yet to the code that invokes returns. The exit() is pre-fork, so it would print logs to stdout/stderr and exit.

Reply:
Sorry, I meant the process that does the killing, not the other way around. I was commenting on the _log.error_ and _log.info_ usage. 
I know it's good to differentiate error level from info level, but sometimes we also want to have things be observable from the data collection side.

Reply:
OK, I'm confused then. I should set the log.info() to log.error()? or the other way around?

Reply:
That's something for you and/or @basepi to decide.

I personally think it's tricky to detect problems from the data if all the relevant details are not there and if all the logs are just the _error_ kind. Think it from the point of view where you have no access to a log file on disk - the only access you have is to the data already sent to a returner.

Reply:
One thing I've been considering is trying to add another custom log level, like `log.splunk` or something, so that we can log things to splunk log handler without having to call them "errors". I'm not sure if it's possible, I'm still a bit of a noob when it comes to the inner workings of the `logging` module.

Reply:
I'm not sure I understand the problem here. I'm going to review the comments, but this all happens before daemonizing, so all the info and error results are printed to console (or systemd logs or whatever) and wouldn't end up in a return.

Except where I send a kill single 15, which would get caught by the usual shutdown signal handler.